Docker SDK unix socket connections blocked

I am using the Docker python SDK, and I get the following error when connections to `127.0.0.1` are allowed, and also passing `--allow-unix-socket`.

> E           docker.errors.DockerException: Error while fetching server API version: A test tried to > use socket.socket.connect() with host "None" (allowed: "127.0.0.1").

I'm using the default `base_url` for the docker client, which is the socket at  `unix:///var/run/docker/docker.sock`, I believe. 

Code reference:
https://github.com/docker/docker-py/blob/b27faa62e792c141a5d20c4acdd240fdac7d282f/docker/api/client.py#L79-L81

I think this is the same issue from #70 / #71.
The combination of `--allow-unix-socket` and `--allow-hosts` does not work as expected.
Take this example test:

```py
import os
import socket
import tempfile

import pytest


def test_unix_connect():
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    with pytest.raises(FileNotFoundError):
        sock.connect('/tmp/socket.sock')


def test_inet_connect():
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    with pytest.raises(ConnectionRefusedError):
        sock.connect(('127.0.0.1', 8080))
 ```
 
 :green_circle:  Running pytest without the plugin will pass (expected):
 ```
$ pytest test.py
 ```
 
:green_circle:  Running pytest with `--disable-socket` will result in two failures (expected):
 ```
$ pytest --disable-socket test.py

FAILED test.py::test_unix_connect - pytest_socket.SocketBlockedError: A test tried to use socket.socket.
FAILED test.py::test_inet_connect - pytest_socket.SocketBlockedError: A test tried to use socket.socket.
```

:green_circle:  Running pytest with `--disable-socket --allow-unix-socket` will result in one failure for the TCP socket use (expected):
```
$ pytest --disable-socket --allow-unix-socket test.py

FAILED example_test.py::test_inet_connect - pytest_socket.SocketBlockedError: A test tried to use socket.socket.
```

:red_circle: Running pytest with `--allow-unix-socket --allow-hosts=127.0.0.1` will result in one failure for the Unix socket use (**unexpected**):
```
$ pytest --allow-unix-socket --allow-hosts=127.0.0.1

FAILED example_test.py::test_unix_connect - pytest_socket.SocketConnectBlockedError: A test tried to use socket.socket.connect() with host "None" (allowed: "127.0.0.1").
```

--- a/pytest_socket.py
+++ b/pytest_socket.py
@@ -56,6 +56,15 @@ def socket_enabled(pytestconfig):
     yield
 
 
+def _is_unix_socket(family) -> bool:
+    try:
+        is_unix_socket = family == socket.AF_UNIX
+    except AttributeError:
+        # AF_UNIX not supported on Windows https://bugs.python.org/issue33408
+        is_unix_socket = False
+    return is_unix_socket
+
+
 def disable_socket(allow_unix_socket=False):
     """ disable socket.socket to disable the Internet. useful in testing.
     """
@@ -63,13 +72,7 @@ def disable_socket(allow_unix_socket=False):
     class GuardedSocket(socket.socket):
         """ socket guard to disable socket creation (from pytest-socket) """
         def __new__(cls, family=-1, type=-1, proto=-1, fileno=None):
-            try:
-                is_unix_socket = family == socket.AF_UNIX
-            except AttributeError:
-                # AF_UNIX not supported on Windows https://bugs.python.org/issue33408
-                is_unix_socket = False
-
-            if is_unix_socket and allow_unix_socket:
+            if _is_unix_socket(family) and allow_unix_socket:
                 return super().__new__(cls, family, type, proto, fileno)
 
             raise SocketBlockedError()
@@ -161,15 +164,7 @@ def socket_allow_hosts(allowed=None, allow_unix_socket=False):
 
     def guarded_connect(inst, *args):
         host = host_from_connect_args(args)
-        if host in allowed:
-            return _true_connect(inst, *args)
-
-        try:
-            is_unix_socket = inst.family == socket.AF_UNIX
-        except AttributeError:
-            # AF_UNIX not supported on Windows https://bugs.python.org/issue33408
-            is_unix_socket = False
-        if allow_unix_socket and is_unix_socket:
+        if host in allowed or (_is_unix_socket(inst.family) and allow_unix_socket):
             return _true_connect(inst, *args)
 
         raise SocketConnectBlockedError(allowed, host)

--- a/pytest_socket.py
+++ b/pytest_socket.py
@@ -130,7 +130,7 @@ def _resolve_allow_hosts(item):
         hosts = mark_restrictions.args[0]
     elif cli_restrictions:
         hosts = cli_restrictions
-    socket_allow_hosts(hosts)
+    socket_allow_hosts(hosts, allow_unix_socket=item.config.__socket_allow_unix_socket)
     return hosts
 
 
@@ -151,7 +151,7 @@ def host_from_connect_args(args):
         return host_from_address(address)
 
 
-def socket_allow_hosts(allowed=None):
+def socket_allow_hosts(allowed=None, allow_unix_socket=False):
     """ disable socket.socket.connect() to disable the Internet. useful in testing.
     """
     if isinstance(allowed, str):
@@ -161,8 +161,17 @@ def socket_allow_hosts(allowed=None):
 
     def guarded_connect(inst, *args):
         host = host_from_connect_args(args)
-        if host and host in allowed:
+        if host in allowed:
             return _true_connect(inst, *args)
+
+        try:
+            is_unix_socket = inst.family == socket.AF_UNIX
+        except AttributeError:
+            # AF_UNIX not supported on Windows https://bugs.python.org/issue33408
+            is_unix_socket = False
+        if allow_unix_socket and is_unix_socket:
+            return _true_connect(inst, *args)
+
         raise SocketConnectBlockedError(allowed, host)
 
     socket.socket.connect = guarded_connect

--- a/tests/test_combinations.py
+++ b/tests/test_combinations.py
@@ -45,3 +45,30 @@ def test_parametrize_with_socket_enabled_and_allow_hosts(testdir, httpbin):
     result.stdout.fnmatch_lines(
         "*SocketConnectBlockedError: A test tried to use socket.socket.connect() with host*"
     )
+
+
+def test_combine_unix_and_allow_hosts(testdir, httpbin):
+    """Test combination of disable, allow-unix and allow-hosts.
+
+    From https://github.com/miketheman/pytest-socket/issues/78
+    """
+    testdir.makepyfile(
+        f"""
+        import socket
+
+        import pytest
+
+
+        def test_unix_connect():
+            sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+            with pytest.raises(FileNotFoundError):
+                sock.connect('/tmp/socket.sock')
+
+
+        def test_inet_connect():
+            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            sock.connect(('{httpbin.host}', {httpbin.port}))
+        """
+    )
+    result = testdir.runpytest("--disable-socket", "--allow-unix-socket", f"--allow-hosts={httpbin.host}")
+    result.assert_outcomes(passed=2)

--- a/tests/test_combinations.py
+++ b/tests/test_combinations.py
@@ -1,4 +1,6 @@
 """Test module to express odd combinations of configurations."""
+# TODO: Move this to a more common location, like `conftest.py`.
+from test_async import unix_sockets_only
 
 
 def test_parametrize_with_socket_enabled_and_allow_hosts(testdir, httpbin):
@@ -47,6 +49,7 @@ def test_parametrize_with_socket_enabled_and_allow_hosts(testdir, httpbin):
     )
 
 
+@unix_sockets_only
 def test_combine_unix_and_allow_hosts(testdir, httpbin):
     """Test combination of disable, allow-unix and allow-hosts.
 

